File: frameworks/domain.core/src/main/java/com/example/frameworks/domain/core/Payload.java
package com.example.frameworks.domain.core;

import static com.example.frameworks.domain.core.MapFieldType.asMap;

public class Payload {
    private final Object value;

    public Payload(Object value) {
        this.value = value;
    }

    public <T> T get(FieldType<T> type) {
        return type.from(value);
    }

    public <T> T get(String key, FieldType<T> type) {
        return type.from(get(asMap()).get(key));
    }
}

File: frameworks/domain.core/src/main/kotlin/com/example/frameworks/domain/core/Payload.kt
package com.example.frameworks.domain.core

import com.example.frameworks.domain.core.MapFieldType.asMap

class Payload(private val value: Any) {
    fun <T> get(type: FieldType<T>): T {
        return type.from(value)
    }

    fun <T> get(key: String, type: FieldType<T>): T {
        return type.from(get(asMap())[key]);
    }
}
